### PR TITLE
leaderElection:v1

### PR DESCRIPTION
**Pull request by zwx14700**
LeaderElection based on prevaga state synchronizer.

### CODE REVIEW

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    public interface LeaderCrashListener {
+        void selectLeader(String name);
+        void crashLeader(String name);
+    }
+
+    public void deleteStream() {
+        try {
+            streamManager.sealStream(scopeName, streamName);
+            Thread.sleep(500);
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: This doesn't look correct. Why is this here?

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+        stateSync = clientFactory.createStateSynchronizer(name,
+                                                    new JavaSerializer<HeartbeatUpdate>(),
+                                                    new JavaSerializer<CreateState>(),
+                                             SynchronizerConfig.builder().build());
+
+        stateSync.initialize(new CreateState(new HashMap<>()));
+    }
+
+    @Data
+    private static class Instance implements Serializable {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: "Instance" is not a good name.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+            return res;
+        }
+
+        /**
+         * When leader is crash we should wait enough time to make sure that the previous leader
+         * is stopping acting as leader.
+         * @param vectorTime The given time.
+         * @return true means has wait enough time, otherwise false.
+         */
+        private boolean isOverWaitTime(long vectorTime) {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: This complicates the algorithm. Rather than this, we should be taking this length of time to detect the death. The host should have an "unhealthy" state which is triggered well before any other host can detect they are dead, where they stop acting as the leader until they can heartbeat.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+            long waitThreshold = vectorTime - WAIT_TO_ELECTION_THRESHOLD * liveInstances.size();
+            //System.out.println("WaiThreshold: " + waitThreshold + " " + vectorTime);
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: Probably remove this. If you want to have debug logging, use a logger and pick an appropriate level.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+    private static abstract class HeartbeatUpdate implements Update<LiveInstances>, Serializable {
+        private static final long serialVersionUID = 1L;
+    }
+
+    @RequiredArgsConstructor
+    private static class HeartBeat extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+        private final Instance temp;
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: Not a good member name.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+    @RequiredArgsConstructor
+    private static class DeclareDead extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+            Map<String, Instance> tempInstances = new HashMap<>(state.liveInstances);
+            tempInstances.remove(name);
+            return new LiveInstances(state.scopedStreamName,
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: If your going to copy liveInstances, a better pattern might be to have a @Builder(toBuilder = true) on LiveInstances then it becomes easier because you can do `state.toBuilder().setXXX(x).build();`

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+                    newRevision,
+                    Collections.unmodifiableMap(tempInstances),
+                    state.vectorTime,
+                    state.leaderName,
+                    state.leaderCrashTime);
+        }
+    }
+
+    @RequiredArgsConstructor
+    private static class ClearLeaderName extends HeartbeatUpdate {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: This shouldn't be needed, we can use the path to declare the host dead.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+                    newRevision,
+                    Collections.unmodifiableMap(state.liveInstances),
+                    state.vectorTime,
+                    null,
+                    crashTime);
+        }
+    }
+
+    @RequiredArgsConstructor
+    private static class LeaderSet extends HeartbeatUpdate {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: This should be called SetLeader

File: build.gradle [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+version '1.0-SNAPSHOT'
+
+sourceCompatibility = 1.8
+
+repositories {
+    mavenLocal()
+    mavenCentral()
+
+}
+
```
Comment (tkaitchuck) [inline build.gradle:38]: Let's add findbugs and checkstyle targets. These can be copied from pravega/pravega

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import io.pravega.client.admin.StreamManager;
+import io.pravega.leaderelection.impl.LeaderElectionImpl;
+import lombok.extern.slf4j.Slf4j;
+import java.net.URI;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+@Slf4j
+public class LeaderElection {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: It is strange to have a concrete class called LeaderElection and another called LeaderElectionImpl. Seeing these file names in the directory it would be easy to guess that this was an interface that the other class implemented.

File: src/test/java/io/pravega/leaderelection/Client.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    public Client(String hostName) {
+        controllerURI = URI.create(DEFAULT_CONTROLLER_URI);
+        le = new LeaderElection(DEFAULT_SCOPE, DEFAULT_CONFIG_NAME, controllerURI,hostName, this);
+        flag = true;
+    }
+
+    @Override
+    public void run() {
+        le.start(500);
+        while(flag);
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/Client.java]: This is an infinate tight loop.

File: src/test/java/io/pravega/leaderelection/Client.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+import lombok.extern.slf4j.Slf4j;
+import java.net.URI;
+
+@Slf4j
+class Client extends Thread implements LeaderElection.LeaderElectionCallback {
+
+    private static final String DEFAULT_SCOPE = "election";
+    private static final String DEFAULT_CONFIG_NAME = "leaderElection";
+    private static final String DEFAULT_CONTROLLER_URI = "tcp://127.0.0.1:9090";
+    private URI controllerURI;
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/Client.java]: Final

File: src/test/java/io/pravega/leaderelection/Client.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    private LeaderElection le;
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/Client.java]: final

File: src/test/java/io/pravega/leaderelection/Client.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    private volatile boolean flag;
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/Client.java]: "flag" is not a good name, it doesn't tell how it is used.
In pravega by convention we use AtomicBoolens

File: src/test/java/io/pravega/leaderelection/LeaderElectionTest.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        assertNotNull(leader);
+
+        List<String> hosts = cfg.getAllConnectHost();
+        hosts.remove(leader);
+        String h1 = hosts.get(0);
+        String h2 = hosts.get(1);
+
+
+        cfg.disconnect(h2);
+        Thread.sleep(1000);
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/LeaderElectionTest.java]: There should not be sleeps in a unit test. Use a latch (See the one in pravega common) or some other way to wake up at the right time.

File: src/test/java/io/pravega/leaderelection/LeaderElectionTest.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        String leader = cfg.checkOneLeader();
+        assertNotNull(leader);
+
+        cfg.disconnect(leader);
+        List<String> hosts = cfg.getAllConnectHost();
+        //disconnect one host
+        cfg.disconnect(hosts.get(0));
+        cfg.disconnect(hosts.get(1));
+        cfg.disconnect(hosts.get(2));
+        Thread.sleep(1000);
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/LeaderElectionTest.java]: Same here and below

File: src/test/java/io/pravega/leaderelection/LeaderElectionTest.java [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.leaderelection;
+import junit.framework.TestCase;
+import org.junit.Test;
+import java.util.List;
+
+
+public class LeaderElectionTest extends TestCase {
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/LeaderElectionTest.java]: Switch to Junit4 style.

File: src/test/java/io/pravega/leaderelection/TestConfig.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+package io.pravega.leaderelection;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+
+public class TestConfig {
+    private Map<String, Client> clientMap;
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/TestConfig.java]: final

File: src/test/java/io/pravega/leaderelection/TestConfig.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    private Set<String> connected;
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/TestConfig.java]: final

File: src/test/java/io/pravega/leaderelection/TestConfig.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+        connected.remove(host);
+        toStop.stopRunning();
+    }
+
+    /**
+     * re-join the server host into the membership.
+     */
+    public void connect(String host) {
+        Client toStart = new Client(host);
+        clientMap.put(host,toStart);
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/TestConfig.java]: This map is not thread safe and is being called from a public method. It is not possible to validate the correctness of this code.

File: src/test/java/io/pravega/leaderelection/TestConfig.java [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+ */
+package io.pravega.leaderelection;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+
+public class TestConfig {
```
Comment (tkaitchuck) [inline src/test/java/io/pravega/leaderelection/TestConfig.java]: This class seems mis named. It is not acting as a config object. Methods like "restartExcept" and "checkOneLeader" and not merely accessors to configuration settings.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 22]
```diff
@@ -0,0 +1,404 @@
...
+            }
+        }
+    }
+
+    public Set<String> getCurrentMembers() {
+        return stateSync.getState().getLiveInstances();
+    }
+
+    public String getCurrentLeader() {
+        return leaderName;
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: This is being updated in the HeartBeater thread and read here in a public method which is by definition a different thread. This is a data race.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 23]
```diff
@@ -0,0 +1,404 @@
...
+    }
+
+    public String getInstanceId() {
+        return instanceId;
+    }
+    
+    public void setRate(int updateRate) {
+        this.updateRate = updateRate;
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: Why is this changable? 
It doesn't look like it will do what you want.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 24]
```diff
@@ -0,0 +1,404 @@
...
+                    state.vectorTime,
+                    state.leaderName);
+        }
+    }
+
+
+    private class HeartBeater implements Runnable {
+        @Override
+        public void run() {
+            try {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: The inside of this try block could be a private method

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 25]
```diff
@@ -0,0 +1,404 @@
...
+            }
+        }
+    }
+
+    /**
+     * Add a new host into the group.
+     */
+    public void add() {
+        stateSync.updateState((state, updates) -> {
+            updates.add(new AddMember(instanceId, state.getVectorTime()));
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: It should check to be sure the instance is not already in the group

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java [hunk 26]
```diff
@@ -0,0 +1,404 @@
...
+            long unhealthyThreshold = vectorTime - UNHEALTHY_THRESHOLD * liveInstances.size();
+            Instance instance = liveInstances.get(name);
+            return instance == null || instance.timestamp >= unhealthyThreshold;
+        }
+        /**
+         * Return all instances that are alived.
+         * @return A set of alive instances.
+         */
+        private Set<String> getLiveInstances() {
+            return liveInstances.keySet();
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: Collections.unmodifyableSet()

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

**Comment by tkaitchuck** [comment]
There should be JavaDocs for all of the public classes and all of their public methods.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java (hunk 1 above)
Comment (zwx14700) [reply src/main/java/io/pravega/leaderelection/LeaderElection.java]: Actually I am not sure whether the leader election should delete the stream. 
I think this should leave to the app, so I don't use this function right now.
When you create leaderelection object, if the scope and stream doesn't exist, it will create stream.
So I wonder if the delete scope and stream should let the leader election to do.

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java (hunk 4 above)
Comment (zwx14700) [reply src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: I have already removed this. It‘s strange why it will still show this

File: src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java (hunk 3 above)
Comment (zwx14700) [reply src/main/java/io/pravega/leaderelection/impl/LeaderElectionImpl.java]: Yes. I have already remove isOverWaitTime a week ago and add healthy check function.

File: src/test/java/io/pravega/leaderelection/LeaderElectionTest.java (hunk 15 above)
Comment (zwx14700) [reply src/test/java/io/pravega/leaderelection/LeaderElectionTest.java]: The purpose of this test is making some members disconnected from group for some time and then rejoin into group. This makes the number of heartbeats of these members less than others. In this way, It make sure we choose the most healthy leader.
disconnect function is not blocking, so we should wait some time then re-connect  to make sure the heartbeats is falling behind.

File: src/test/java/io/pravega/leaderelection/TestConfig.java (hunk 20 above)
Comment (zwx14700) [reply src/test/java/io/pravega/leaderelection/TestConfig.java]: The connect and disconnect only operated by the Main thread.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 27]
```diff
@@ -0,0 +1,493 @@
...
+    @RequiredArgsConstructor
+    private static class HeartBeat extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+        private final InstanceInfo instanceInfo;
+
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+            Map<String, InstanceInfo> tempInstances = new HashMap<>(state.liveInstances);
+            long vectorTime = Long.max(tempInstances.values().stream().max(InstanceInfo::compare).get().timestamp, instanceInfo.timestamp);
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: This could be done more cleanly by first mapping the instance to the timestamp. 
Then you could use the builtin for Max, and not need the compare and the result would be the timestamp.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 28]
```diff
@@ -0,0 +1,493 @@
...
+     * 1. update timestamp and timeout.
+     * 2. remove all dead instances.
+     * 3. If leader dead or not exist, select a new one.
+     */
+    private void sendHeartbeat() {
+        stateSync.updateState((state, updates) -> {
+            long vectorTime = state.getVectorTime() + 1;
+            InstanceInfo instanceInfo = state.liveInstances.get(instanceId);
+            long newTimes = instanceInfo.times + 1;
+            double newTimeout = (instanceInfo.timeout * instanceInfo.times +
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: Here and below: Feel free to take advantage of a larger line width. In Pravega we use 120.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 29]
```diff
@@ -0,0 +1,493 @@
...
+
+    private class HeartBeater implements Runnable {
+        @Override
+        public void run() {
+            try {
+                stateSync.fetchUpdates();
+                notifyListener();
+                sendHeartbeat();
+                // when leader changes, notify to all.
+                if (leaderName == null || !leaderName.equals(stateSync.getState().leaderName)) {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: This is comparing and changing leaderName without any threadsafe check. You probably need a synchronized block. My advice would be to make a tiny private class that just manages the leader variable. that way the outer class won't have to think about it.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 30]
```diff
@@ -0,0 +1,493 @@
...
+    }
+
+    private void setRate(int updateRate) {
+        this.updateRate = updateRate;
+    }
+
+    private void add() {
+        if (!stateSync.getState().liveInstances.containsKey(instanceId)) {
+            stateSync.updateState((state, updates) -> {
+                updates.add(new AddMember(instanceId, state.getVectorTime()));
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java:373]: Make sure the instance is not already there.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 31]
```diff
@@ -0,0 +1,493 @@
...
+                    listener.stopActingLeader();
+                }
+            }
+        }
+    }
+
+    /**
+     * The interface of LeaderElection callback.
+     */
+    public interface LeaderElectionCallback {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: This should probably be in its own file.

File: src/test/java/io/pravega/leaderelection/LeaderElectionTest.java (hunk 15 above)
Comment (tkaitchuck) [reply src/test/java/io/pravega/leaderelection/LeaderElectionTest.java]: Sure. makes sense. So let's add a hook inside the impl so that test code can get woken up every time the state updates. Then block on that.

**Review by tkaitchuck — COMMENTED** [review]
It's probably worth doing a randomized test. 
Like maybe have Instances come and go and then after some time assert there is a leader and only the live instances are in the state, etc.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java (hunk 29 above)
Comment (zwx14700) [reply src/main/java/io/pravega/leaderelection/LeaderElection.java]: The leadername is a local copy of the leadername in state synchronizer. I think there is no need to do threadsafe check. only the heartbeater thread can check and change the local leadername.

File: src/main/java/io/pravega/leaderelection/Leader.java [hunk 32]
```diff
@@ -0,0 +1,28 @@
+package io.pravega.leaderelection;
+
+public class Leader {
+    private String leaderName;
+    private Leader() {
+        leaderName = null;
+    }
+
+    private static class LeaderSingleton {
+        private static final Leader INSTANCE = new Leader();
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/Leader.java]: This is introducing global state. This will break users in crazy and unexpected ways. Consider: It might be possible for an application to be running two different leader electors in the same process. Perhaps one is in a library. Or perhaps there are multiple unit tests for leader election, and we turn on parallelism to speed up our build. Then suddenly they interfere with one another and start setting incorrect state here.

What's more this is not thread safe, so it could be even more confusing.

File: src/main/java/io/pravega/leaderelection/Leader.java [hunk 33]
```diff
@@ -0,0 +1,28 @@
+package io.pravega.leaderelection;
+
+public class Leader {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/Leader.java]: If this class is public it should be thread safe and be fully documented.

File: src/main/java/io/pravega/leaderelection/Leader.java [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+
+    public String getLeader() {
+        return leaderName;
+    }
+
+    public void setLeader(String name) {
+        leaderName = name;
+    }
+
+    public Boolean isLeader(String name) {
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/Leader.java]: boolean

File: src/main/java/io/pravega/leaderelection/Leader.java (hunk 32 above)
Comment (zwx14700) [reply src/main/java/io/pravega/leaderelection/Leader.java]: Yes. Changing to this way is not a good way. Maybe I could take leader into a private static class in the leader election.

File: src/main/java/io/pravega/leaderelection/LeaderElection.java [hunk 35]
```diff
@@ -0,0 +1,491 @@
...
+                                                    new JavaSerializer<CreateState>(),
+                                             SynchronizerConfig.builder().build());
+
+        stateSync.initialize(new CreateState(new HashMap<>()));
+    }
+
+    @Data
+    private static class InstanceInfo implements Serializable {
+        private final long timestamp;
+        private final long times;
```
Comment (tkaitchuck) [inline src/main/java/io/pravega/leaderelection/LeaderElection.java]: This needs a better name

**Review by tkaitchuck — APPROVED** [review]
